### PR TITLE
Reconnection fuzz tests

### PR DESCRIPTION
**Pull request by daesun-park**
## Description

This PR adds reconnect/disconnect ops to the sharedTree fuzz tester.

### CODE REVIEW

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts
Comment (justus-camp) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:475]: Can we use `ConnectionState` from `container-definitions` instead of hardcoding the numerical values? Would make this more readable.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts
Comment (justus-camp) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:396]: This is unnecessarily confusing. Can just `return containersInfo !== undefined`.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts
Comment (daesun-park) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts]: updated to use `ConnectionState` enum

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts
Comment (daesun-park) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts]: simplified to `return containersInfo !== undefined`

**Review by justus-camp — APPROVED** [review]

File: experimental/dds/tree2/src/test/shared-tree/fuzz/operationTypes.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/operationTypes.ts:116]: Why is this index needed here? Can't we use the index from `TreeSelectionContext`? Same for `ReconnectOp.index`.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/topLevelFuzz.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/topLevelFuzz.spec.ts:34]: Why this change?

File: experimental/dds/tree2/src/test/shared-tree/fuzz/operationTypes.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/operationTypes.ts:117]: This field doesn't seem to be consumed anywhere. Same for `ReconnectOp.isObserver`.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts:65]: Nit: flipping these two lines would let you remove the `?`:

```suggestion
			assert(state.testTreeProvider !== undefined);
			state.testTreeProvider.opProcessingController.resumeProcessing();
```

**Review by yann-achard-MS — COMMENTED** [review]
I left some comments. I think it makes sense to wait for Abram before attempting to merge this in.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts
Comment (Abe27342) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts:93]: this isn't actually the reconnect flow--it looks like it's triggering offline load instead. I'm not sure we have an easy way to incur the reconnect/resubmit flow when using the e2e infra like we've done here. You might be able to do it by forcing connection state change (e.g. using `forceReadOnly`) combined with pausing/resuming the delta connection as you've done, but I'm not sure it's the best approach.

Might be easiest to chat about alternatives in-person... I have a few ideas but depends somewhat on the goals.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:393]: fwiw--the DDS fuzz harness gets around this by making a single `changeConnectionState` op type with a value of what the new state should be, and always flips it. this doesn't let you customize relative probabilities of disconnect vs. reconnect, but I don't think that's a very interesting feature anyway. You probably don't want your clients to 'tend to disconnect' over time, and if reconnect probability is higher than disconnect probability you're less likely to have multiple clients disconnected at once which can be interesting.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts
Comment (justus-camp) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts:93]: I had the same thought about this not really being the reconnect flow, but figured that the goal was to test stashed ops/offline load. Is that the intent or should we be doing that separately and adding offline load separately?

Comment (Abe27342) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditReducers.ts:93]: ideally we test both. it's somewhat awkward that our e2e test infra doesn't allow easy testing of the reconnect+resubmit flow, whereas our mocks don't allow easy testing of the offline load flow. but i'm not sure it's super important that we test them together as part of the same fuzz test.

I'm scheduled for a brown-bag (which keeps getting pushed back due to other meetings taking my timeslot :P) to talk about a generic DDS fuzzing harness I wrote (see work item [here](https://dev.azure.com/fluidframework/internal/_workitems/edit/2280)) that's now ready to use (and is being used in 3-4 other DDSes) and handles reconnect out-of-the-box. It'd be great if SharedTree's fuzz testing was set up to plug nicely into that--it doesn't have to be *all* of the fuzz tests that you write, but it would be one way to work around this issue. I have a lot of opinions in this area though, so as I said might be easier to just chat about it in person.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:415]:
```suggestion
	return connectionOpGenerator;
```
should suffice actually. or you could just inline the return on the `createWeightedGenerator` line.

Comment (Abe27342) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:415]: also.... @yann-achard-MS how did you make the suggestion apply to multiple lines? I thought GH didn't support that all this time

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:415]: You do mouse-button-down on the top line, move your mouse cursor down to the last line, and do mouse-button-up. I had to google it :)

Comment (Abe27342) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts:415]: 🥇

**Pull request closed without merging**
